src/pe/debug.rs

```diff
@@ -36,88 +163,176 @@ impl<'a> DebugData<'a> {
         file_alignment: u32,
         opts: &options::ParseOptions,
     ) -> error::Result<Self> {
-        let image_debug_directory =
-            ImageDebugDirectory::parse_with_opts(bytes, dd, sections, file_alignment, opts)?;
-        let codeview_pdb70_debug_info =
-            CodeviewPDB70DebugInfo::parse_with_opts(bytes, &image_debug_directory, opts)?;
-        let codeview_pdb20_debug_info =
-            CodeviewPDB20DebugInfo::parse_with_opts(bytes, &image_debug_directory, opts)?;
+        Self::parse_with_opts_and_fixup(bytes, dd, sections, file_alignment, opts, 0)
+    }
+
+    pub fn parse_with_opts_and_fixup(
+        bytes: &'a [u8],
+        dd: data_directories::DataDirectory,
+        sections: &[section_table::SectionTable],
+        file_alignment: u32,
+        opts: &options::ParseOptions,
+        rva_offset: u32,
+    ) -> error::Result<Self> {
+        let offset =
+            utils::find_offset(dd.virtual_address as usize, sections, file_alignment, opts)
+                .ok_or_else(|| {
+                    error::Error::Malformed(format!(
+                        "Cannot map ImageDebugDirectory rva {:#x} into offset",
+                        dd.virtual_address
+                    ))
+                })?;
+
+        // Ensure that the offset and size do not exceed the length of the bytes slice
+        if offset + dd.size as usize > bytes.len() {
+            return Err(error::Error::Malformed(format!(
+                "ImageDebugDirectory offset {:#x} and size {:#x} exceeds the bounds of the bytes size {:#x}",
+                offset, dd.size, bytes.len()
+            )));
+        }
+        let data = &bytes[offset..offset + dd.size as usize];
```

Comment:
probably in the future we should port this kind of code to something like:

```rust
let Some (data) = bytes.get(offset..offset + dd.size as usize) else {
              return Err(error::Error::Malformed(format!(
                "ImageDebugDirectory offset {:#x} and size {:#x} exceeds the bounds of the bytes size {:#x}",
                offset, dd.size, bytes.len()
            )));
};
```

this is clearer, and doesn't repeat the bounds, but we its minor nit. I should have remembered this earlier, as we've done these in several other places, but the `else` guard syntax only merged fairly recently (maybe a year or so ago iirc, but maybe even longer :sweat_smile: )